docs/RingBuffer.md

```diff
@@ -0,0 +1,278 @@
+# New ebpf Ring Buffer Map (proposal)
+
+## Overview
+
+The current ringbuffer uses a pure callback-based approach to reading the ringbuffer.
+Linux also supports memory-mapped polling consumers, which can't be directly supported in the current model.
+
+The new API will support 2 consumer types: callbacks and direct access to the mapped producer memory (with poll to wait for data).
+
+Callback consumer:
+
+1. Call `ring_buffer__new` to set up callback.
+2. The callback will be invoked for each record written to the ring buffer.
+
+Mapped memory consumer:
+
+1. Call `ebpf_ring_buffer_get_buffer` to get pointers to the mapped producer/consumer pages.
+2. Call `ebpf_ring_buffer_get_wait_handle` to get the wait handle.
+3. Directly read records from the producer pages (and update consumer offset as we read).
+4. Call `WaitForSingleObject`/`WaitForMultipleObject` as needed to wait for new data to be available.
```

Comment:
Line 6 says Linux supports polling consumers.  What APIs do callers use on Linux to get this behavior?
Why can't those APIs work on Windows?  Why are new `ebpf_` APIs needed?

This document doesn't appear to answer any of these questions.

Reply:
Updated to discuss the the linux APIs. On linux mmap is used for mapping the memory and epoll used for waiting. There aren't functionally equivalent APIs on Windows, so I also added a windows-specific function to get the wait handle (instead of `ring_buffer__epoll_fd()`, since we don't return an epoll fd).

Reply:
For `mmap`, bing copilot says it can be done if the driver uses `ZwCreateSection`, and the user process does `MapViewOfFile` on the result.  I didn't ask it about epoll.

But could the new apis be implemented on top of the Linux mmap and epoll?  If so, then they're technically not windows specific, right?

Reply:
Yes, get_buffer and get_wait_handle could be implemented on top of `mmap` and `ring_buffer_epoll_fd` (with an intermediate translation layer), so they aren't necessarily Windows specific.

On linux, those 2 parts of the ringbuffer API are already linux specific though -- they use the returned values directly with linux syscalls to map the memory and wait on the epoll fd.

Redesigning the libbpf API to be more platform independent is possible, but this proposal keeps the rest of the linux libbpf API as-is and just replaces the 2 linux-specific parts with windows-specific functions, plus adds asyc/sync options since linux is synchronous-only. To avoid a compatibility break for existing Windows customers, the default behavior on Windows is also different from Linux.

That redesign effort seems like a separate task, and when that happens, we could replace/wrap the Windows functions with the generic ones.

---

On creating a translation layer to hide the Windows-specific parts:

ZwCreateSection is limited to memory mappings backed by files or the page file, so it is not directly compatible with the ringbuffer mmap usage on linux, and I don't think there are directly compatible APIs to hide the functionality behind something that looks like mmap.

Similarly, epoll's behavior and design is not quite the same as KEVENT. There are some epoll-like APIs implemented in Windows, and we could implement `epoll_fd` to return one of those wrapped in an fd, but we opted not to do that to avoid hiding the differences between the designs, and for it to be intuitive/obvious which other Windows APIs the ring buffer wait handle works with (which on Linux is already the case because the API is linux-specific).